internal/manager/conditions_crd.go

```diff
@@ -0,0 +1,43 @@
+package manager
+
+import (
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/controllers/utils"
+)
+
+// CRDControllerCondition determines if a resource controller for a given CRD should be enabled.
+type CRDControllerCondition struct {
+	gvr           schema.GroupVersionResource
+	toggleEnabled bool
```

Comment:
My point was to embed this logic in a single place as it's the same for every CRD controller. It makes it easily unit testable and it's not possible to omit the toggle in case someone forgets about it. 

Also, passing the toggle to the `CRDControllerCondition` makes it clearly visible that the `disabling the '%s' controller due to missing CRD installation` log will be visible only when the toggle is on.

Reply:
I still think it is weird to pass a `false` (from calculation on other conditions) to `toggleEnabled` to disable the controller, instead of using `&&` expressions. Because the logic made the controller to do judgement that it does not aware of.

> Also, passing the toggle to the CRDControllerCondition makes it clearly visible that disabling the '%s' controller due to missing CRD installation will be logged only when the toggle is on.

This does not make sense. Expression `someCondition && crdCondition.Enabled()` uses short circuit evaluation, so that if `someCondition` is false, `crdCondition.Enabled()` is not called, then the log is not printed. So if we do not want the log to appear when other conditions are not satisfied, just put the `.Enabled()` at the last of `&&` parts.

Reply:
I don't think it should go into this constructor method. Sure, we can have other function, but embedding that into constructor is misleading. By the way, we don't have a single place where object returned by `NewCRDCondition()` is used other than `NewCRDCondition(...).Enabled()`. So I wonder if that shouldn't be a standalone function

Reply:
Let's make it simple then: [5e546bf](https://github.com/Kong/kubernetes-ingress-controller/pull/3013/commits/5e546bfd26ff3ede34e066d01b426805f45ed35b) && [9080e0f](https://github.com/Kong/kubernetes-ingress-controller/pull/3013/commits/9080e0fc5f53c611526ff1d2445d198304b3058e)